### ISSUE TITLE
cKDTree INF bug

### ISSUE THREAD

**Issue by SteveDoyle2** [scipy/scipy#6425]
I'm running into a problem with cKDTree with scipy 0.18.0 on both Windows and Linux.  It doesn't calculate the expected query like KDTree does.  It seems to behave consistently across Python versions as well.

I've tested this on scipy 0.18.0 using Python 2.7, 3.4, and 3.5.  Scipy 0.17.1 and 0.16.0 work with no problems.

Any ideas?

    import numpy as np
    import scipy
    import scipy.spatial

    xyz = np.array([
        [ 0.,   0.,   0.],
        [ 1.01, 0.,   0.],
        [ 0.,   1.,   0.],
        [ 0.,   1.01, 0.],
        [ 1.,   0.,   0.],
        [ 1.,   1.,   0.],],
    dtype='float32')

    ckdt = scipy.spatial.cKDTree(xyz)

    deq, ieq = ckdt.query(xyz, k=4, distance_upper_bound=0.2)
    print(deq)

    #>>> [[ inf  inf  inf  inf]
    # [ inf  inf  inf  inf]
    # [ inf  inf  inf  inf]
    # [ inf  inf  inf  inf]
    # [ inf  inf  inf  inf]
    # [ inf  inf  inf  inf]]

    kdt = scipy.spatial.KDTree(xyz)
    deq2, ieq2 = kdt.query(xyz, k=4, distance_upper_bound=0.2)
    print(deq2)

    #>>> [[ 0.                 inf         inf         inf]
    # [ 0.          0.00999999         inf         inf]
    # [ 0.          0.00999999         inf         inf]
    # [ 0.          0.00999999         inf         inf]
    # [ 0.          0.00999999         inf         inf]
    # [ 0.                 inf         inf         inf]]

**Comment by ev-br** [comment]
@sturlamolden @rainwoodman thoughts?

**Comment by sturlamolden** [comment]
I think the bug might be on line 690 in ckdtree.pyx (missing dtype qualifier), but I cannot fathom why it did not crash on line 702 as it should when dtype of array xx is incorrect.

**Comment by rainwoodman** [comment]
1. changing dtype to float64 doesn't solve this.
2. increasing upper_bound to 2.0 resolves the issue, partially:

In [11]: deq, ieq = ckdt.query(xyz, k=4, distance_upper_bound=3)

In [12]: print(deq)
[[  0.00000000e+00   1.00000000e+00   1.00000000e+00   1.00999999e+00]
 [  9.53674317e-09   1.00000000e-02   1.00005000e+00   1.01000000e+00]
 [  0.00000000e+00   9.99999046e-03   1.00000000e+00   1.00000000e+00]
 [  9.53674317e-09   1.00000000e-02   1.00005000e+00   1.01000000e+00]
 [  0.00000000e+00   9.99999046e-03   1.00000000e+00   1.00000000e+00]
 [  0.00000000e+00   1.00000000e+00   1.00000000e+00   1.00005000e+00]]

I don't see why numbers are so inaccurate. could be upper-bound implies some approximation?

**Comment by rainwoodman** [comment]
3. This does not affect sparse distance matrix calculation.
```
In [16]: ckdt.sparse_distance_matrix(ckdt, 0.2).todense()
Out[16]: 
matrix([[ 0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
          0.        ],
        [ 0.        ,  0.        ,  0.        ,  0.        ,  0.00999999,
          0.        ],
        [ 0.        ,  0.        ,  0.        ,  0.00999999,  0.        ,
          0.        ],
        [ 0.        ,  0.        ,  0.00999999,  0.        ,  0.        ,
          0.        ],
        [ 0.        ,  0.00999999,  0.        ,  0.        ,  0.        ,
          0.        ],
        [ 0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
          0.        ]])
```

**Comment by rainwoodman** [comment]
It is  line 470 of query.cxx.

 ```  
  /* fill output arrays with sorted neighbors */
    int j = nk - 1;
    for (i=neighbors.n-1; i>=0; --i) {
        neighbor = neighbors.pop();

        if(i + 1 != k[j]) continue;
        result_indices[j] = neighbor.contents.intdata;
        if (NPY_LIKELY(p == 2.0))
            result_distances[j] = std::sqrt(-neighbor.priority);
        else if ((p == 1.) || (ckdtree_isinf(p)))
            result_distances[j] = -neighbor.priority;
        else
            result_distances[j] = std::pow((-neighbor.priority),(1./p));
        --j;
    }
```

I don't know what the if condition is for, but apparently it is skipping all outputs when the index is not full.. Maybe @sturlamolden can comment on this. Seems to be a mismatch between indices.

**Comment by rainwoodman** [comment]
Found it!

**Comment by rainwoodman** [comment]
Fixed in #6431 .
